android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -2925,7 +2925,17 @@ private static void performLeftDeletions(InputConnection ic, int dn) {
         return;
       }
 
-      int numPairs = CharSequenceUtil.countSurrogatePairs(charsBackup, dn);
+      // Count the number of characters which are surrogate pairs
+      int index = lastIndex, dnx = dn, numPairs = 0;
+      while(index > 0 && dnx > 0) {
+        if(Character.isLowSurrogate(charsBackup.charAt(index)) &&
+            Character.isHighSurrogate(charsBackup.charAt(index-1))) {
+          numPairs++;
+          index--;
+        }
+        index--;
+        dnx--;
```

Comment:
I read the commit message that countSurrogatePairs() was incorrect for system keyboard, but that it was used elsewhere correctly.

Should we document the limitation in CharSequenceUtil.countSurrogatePairs()? (I'm also still trying to understand the difference)

Reply:
`CharSequenceUtil.countSurrogatePairs()` counts the number of surrogate pairs in a string of `n` code units (i.e. 16 bit chars) length. This is counting the number of surrogate pairs in a string of `n` code points (which may be either single 16 bit chars or surrogate pairs).

Reply:
I still don't get why `CharSequenceUtil.countSurrogatePairs()`  is considered to be working fine? It didn't work which is why it was replaced with a inline solution so how can we have confidence it is working?

Reply:
Continuing the `countSurrogatePairs()` discussion in #6301. Needs review but I want to try and get this selection stuff sorted.